src/backend/distributed/executor/adaptive_executor.c

```diff
@@ -945,7 +1011,11 @@ StartDistributedExecution(DistributedExecution *execution)
 	 * If the current or previous execution in the current transaction requires
 	 * rollback then we should use transaction blocks.
 	 */
-	execution->isTransaction = InCoordinatedTransaction();
+	if (InCoordinatedTransaction() && execution->useRemoteTransactionBlocks ==
+		REMOTE_TRANSACTION_BLOCKS_ALLOWED)
```

Comment:
what if it is DISALLOWED?

Reply:
we have three states:
- by default a transaction can be opened so it is **allowed**.
- If there are dependent jobs then a transaction should not be opened, so it is **disallowed**.
- If we are in a coordinated transaction and there are no dependent jobs, then we should open a transaction, so it is **required**. 

If we are in a coordinated transaction but we have dependent jobs then we do not open transaction so that is why the check is here.

I will add it as a comment here.

Reply:
I meant:

What if we are in a coordinated transaction and useRemoteTransactionBlocks is DISALLOWED? We should throw an error or have an assert.

Reply:
when we have a query like:

```sql
begin;
a repartition query;
end;
```

do we want to throw an error? I dont think we do.

Reply:
but `InCoordinatedTransaction()` wouldn't be true in that case, since `DistributedExecutionRequiresRollback` returns false for the repartition query. 

This code is pretty confusing to me. The outcome of DistributedExecutionRequiresRollback depends on the value of useRemoteTransactionBlocks, but now the value of useRemoteTransactionBlocks depends on the outcome of DistributedExecutionRequiresRollback.

Reply:
In `InCoordinatedTransaction()` we check if the transaction state is NONE or IDLE.

When we have a subquery we start a transaction so the transaction state will be STARTED, after that if we have a repartition query in the subquery then the DistributedExecutionRequiresRollback will return false because this is a repartition query however our transaction state is STARTED, so in that case `InCoordinatedTransaction()` would return true. 

I have moved the decision of whether to open a transaction or not to the creation of execution.